src/common/SharedMemory.cpp

```diff
@@ -0,0 +1,230 @@
+/*
+ * SharedMemory.cpp
+ *
+ * Copyright (c) 2022 Dominic Clark <mrdomclark/at/gmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ */
+
+#include "SharedMemory.h"
+
+#include "lmmsconfig.h"
+
+#ifdef LMMS_HAVE_UNISTD_H
+#	include <unistd.h>
+#endif
+
+#if _POSIX_SHARED_MEMORY_OBJECTS > 0
+#	include <system_error>
+#
+#	include <sys/mman.h>
+#	include <sys/stat.h>
+#	include <assert.h>
+#	include <fcntl.h>
+#else
+#	include <stdexcept>
+#
+#	include <QtGlobal>
+#	include <QSharedMemory>
+#endif
+
+namespace detail {
+
+#if _POSIX_SHARED_MEMORY_OBJECTS > 0
+
+namespace {
+
+template<typename F>
+int retryWhileInterrupted(F&& function) noexcept(std::is_nothrow_invocable_v<F>)
+{
+	int result;
+	do
+	{
+		result = function();
+	}
+	while (result == -1 && errno == EINTR);
+	return result;
+}
+
+template<typename T, T Null>
+class NullableResource
```

Comment:
This class looks like a good candidate to be contained in a separate header file for future uses.

Reply:
Good idea, will move it to a new header `include/RaiiHelpers.h`.